impl/src/main/java/io/kroxylicious/testing/kafka/common/KafkaTopology.java

```diff
@@ -0,0 +1,237 @@
+ * 3. manifests the topology using the driver, creating the nodes and tracking their state
+ * 4. enables manipulation of the topology, stopping/starting/removing nodes.
+ */
+public class KafkaTopology {
+    private final KafkaDriver driver;
+    private final KafkaClusterConfig config;
+
+    // kludge: creating a KafkaNode may consult the NodeConfigurations of the Topology looking for controllers, so we
+    // need the cluster configuration to be available and associated with the topoloogy for all nodes before we
+    // construct any KafkaNodes
+    private final Map<Integer, KafkaNodeConfiguration> nodeConfigurations;
+    private final Map<Integer, KafkaNode> nodes;
+    private final ZookeeperConfig zookeeperConfig;
+
+    private static final Comparator<KafkaNode> NON_CONTROLLERS_FIRST = (o1, o2) -> {
+        if (!o1.configuration().isController()) {
+            return -1;
+        }
+        else if (!o2.configuration().isController()) {
+            return 1;
+        }
+        else {
+            return 0;
+        }
+    };
+    private final Zookeeper zookeeper;
+
+    private KafkaTopology(KafkaDriver driver, KafkaClusterConfig config) {
+        this.driver = driver;
+        this.config = config;
+        if (!config.isKraftMode()) {
+            this.zookeeperConfig = new ZookeeperConfig(driver);
+            this.zookeeper = driver.createZookeeper(zookeeperConfig);
+        }
+        else {
+            this.zookeeperConfig = null;
+            this.zookeeper = null;
+        }
+        this.nodes = new HashMap<>();
+        this.nodeConfigurations = new HashMap<>();
+    }
+
+    boolean isKraftMode() {
+        return this.config.isKraftMode();
+    }
+
+    public static KafkaTopology create(KafkaDriver driver, KafkaClusterConfig config) {
+        KafkaTopology kafkaTopology = new KafkaTopology(driver, config);
+        Map<Integer, KafkaNodeConfiguration> configs = generateConfigurations(driver, config, kafkaTopology);
+        kafkaTopology.nodeConfigurations.putAll(configs);
+        // createNode requires the topology to have all nodes configurations installed first, so it can ask for all
+        // controller node details.
+        Map<Integer, KafkaNode> nodes = configs.values().stream()
+                .collect(Collectors.toMap(KafkaNodeConfiguration::nodeId, driver::createNode));
+        kafkaTopology.addNodes(nodes);
+        return kafkaTopology;
+    }
+
+    @NotNull
+    private static Map<Integer, KafkaNodeConfiguration> generateConfigurations(KafkaDriver driver, KafkaClusterConfig config, KafkaTopology kafkaTopology) {
+        int nodesToCreate = Math.max(config.getBrokersNum(), config.getKraftControllers());
+        Stream<KafkaNodeConfiguration> nodeConfigurationStream = IntStream.range(0, nodesToCreate)
+                .mapToObj(value -> new IdAndRoles(roles(config, value), value))
+                // note that generation of a configuration may allocate resources like ports to it
```

Comment:
I feel like this comment should be in the javadoc for the method or at call sites such as line 75.

Or is it meant to convey that `getNodeConfiguration` might do the allocation? In which case this maybe the right location for the comment but `getNodeConfiguration` should probably be re-named